File: backend/warranty/src/main/java/edu/uth/warranty/model/Report.java
package edu.uth.warranty.model;

import jakarta.persistence.*;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.math.BigDecimal;

import lombok.Getter;
import lombok.Setter;   
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;


@Entity
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Table(name = "Report")
public class Report {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long reportId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "claim_id", nullable = false)
    private WarrantyClaim claim;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "technician_id", nullable = false)
    private Technician technician;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "service_center_id", nullable = false)
    private ServiceCenter center;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "vehicle_id", nullable = false)
    private Vehicle vehicle;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "recall_campaign_id", nullable = false)
    private RecallCampaign campaign;

    @Column(name = "status", nullable = false)
    private String status;
    @Column(name = "report_date", nullable = false)
    private LocalDate reportDate;

    @Column(name = "started_at")
    private LocalDateTime startedAt;

    @Column(name = "finished_at")
    private LocalDateTime finishedAt;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(columnDefinition = "TEXT")
    private String actionTaken;

    @Column(columnDefinition = "TEXT")
    private String partUsed;

    @Column(columnDefinition = "TEXT")
    private String replacedPart;

    @Column(name = "part_cost", precision = 10, scale = 2)
    private BigDecimal partCost;

    @Column(name = "labor_cost", precision = 10, scale = 2)
    private BigDecimal actualCost;

    @Column(name = "created_date")
    private LocalDateTime createdDate;

    @Column(name = "created_by")
    private String createdByUsername;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "updated_by")
    private String updatedBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false) // FK trỏ về User
    private User createdBy;
}

File: frontend/src/types/report.ts
// frontend/src/types/report.ts
export interface ReportRequest {
    id?: number;
    claimId: number;
    technicianId: number;
    centerId: number;
    vehicleId: number;
    campaignId?: number;
    
    createdById: number; 

    status: string;
    reportDate?: string; 

    startedAt?: string; 
    finishedAt?: string; 

    description?: string;
    actionTaken?: string;
    partUsed?: string;
    replacedPart?: string;

    partCost: number;      
    actualCost: number;   
    
    createdByText?: string; 
    updatedBy?: string;
    diagnosticCodes?: string;
}

export interface ReportResponse {
    id: number;

    status: string; 
    reportDate: string;
    diagnosticCodes?: string; 
    
    partCost: number;
    actualCost: number;
    totalCalculatedCost: number; 

    startedAt: string;
    finishedAt: string;
    actionTaken: string;
    
    claimId: number;
    technicianId: number;
    technicianName: string;
    centerId: number;
    vehicleId?: number;
    campaignId?: number;
    
    createdByText: string;
    createdDate: string;
    updatedAt: string;

    description?: string;
    
    partUsed?: string;
    replacedPart?: string;

}
